applications/EigenSolversApplication/custom_solvers/feast_eigensystem_solver.h

```diff
@@ -0,0 +1,435 @@
+/*
+//  KRATOS _______
+//        / ____(_)___ ____  ____
+//       / __/ / / __ `/ _ \/ __ \
+//      / /___/ / /_/ /  __/ / / /
+//     /_____/_/\__, /\___/_/ /_/ SolversApplication
+//             /____/
+//
+//  Author:  Quirin Aumann
+*/
+
+#if !defined(KRATOS_FEAST_EIGENSYSTEM_SOLVER_H_INCLUDED)
+#define KRATOS_FEAST_EIGENSYSTEM_SOLVER_H_INCLUDED
+
+// External includes
+
+// Project includes
+#include "includes/define.h"
+#include "includes/kratos_parameters.h"
+#include "linear_solvers/linear_solver.h"
+#include "includes/ublas_interface.h"
+#include "includes/ublas_complex_interface.h"
+
+extern "C" {
+    #include <feast.h>
+    #include <feast_sparse.h>
+}
+
+namespace Kratos
+{
+
+template<
+    bool TSymmetric,
+    typename TScalarIn,
+    typename TScalarOut,
+    class TSparseSpaceTypeIn = TUblasSparseSpace<TScalarIn>,
+    class TDenseSpaceTypeIn = TUblasDenseSpace<TScalarIn>,
+    class TSparseSpaceTypeOut = TUblasSparseSpace<TScalarOut>,
+    class TDenseSpaceTypeOut = TUblasDenseSpace<TScalarOut>>
+class FEASTEigensystemSolver
+    : public LinearSolver<TSparseSpaceTypeIn, TDenseSpaceTypeIn>
+{
+    Parameters mParam;
+
+  public:
+    KRATOS_CLASS_POINTER_DEFINITION(FEASTEigensystemSolver);
+
+    typedef LinearSolver<TSparseSpaceTypeIn, TDenseSpaceTypeIn> BaseType;
+
+    typedef typename TSparseSpaceTypeIn::MatrixType SparseMatrixType;
+
+    typedef typename TSparseSpaceTypeIn::VectorType VectorType;
+
+    typedef typename TDenseSpaceTypeIn::MatrixType DenseMatrixType;
+
+    typedef typename TDenseSpaceTypeOut::MatrixType DenseOutputMatrixType;
+
+    typedef matrix<TScalarOut, column_major> FEASTMatrixType;
+
+    typedef typename TSparseSpaceTypeOut::VectorType OutputVectorType;
+
+    typedef TScalarIn ValueType;
+
+    FEASTEigensystemSolver(
+        Parameters param
+    ) : mParam(param)
+    {
+        Parameters default_params(R"(
+        {
+            "solver_type" : "eigen_feast",
+            "symmetric" : true,
+            "number_of_eigenvalues" : 0,
+            "search_lowest_eigenvalues" : false,
+            "search_highest_eigenvalues" : false,
+            "e_min" : 0.0,
+            "e_max" : 0.0,
+            "e_mid_re" : 0.0,
+            "e_mid_im" : 0.0,
+            "e_r" : 0.0,
+            "subspace_size" : 0,
+            "max_iteration" : 20,
+            "tolerance" : 1e-12,
+            "echo_level" : 0
+        })");
+
+        mParam.ValidateAndAssignDefaults(default_params);
+
+        KRATOS_ERROR_IF( mParam["number_of_eigenvalues"].GetInt() < 0 ) <<
```

Comment:
is 0 valid input?

```suggestion
        KRATOS_ERROR_IF( mParam["number_of_eigenvalues"].GetInt() < 1 ) <<
```

Reply:
yes. If a subspace size is specified, the number of eigenvalues is ignored (it only provides a rough estimate on the subspace size). I will add a few more checks though